src/device/device_ledger.cpp

```diff
@@ -90,6 +90,18 @@ namespace hw {
       AKout = keys.AKout;
     }
 
+    ABPkeys &ABPkeys::operator=(const ABPkeys& keys) {
```

Comment:
Does everything get copied? Because this looks like a `ABPkeys operator=(const ABPkeys&) = default;` case (the copy constructor too).

If this style is preferred since the copy constructor is already explicit:
```c++
if (std::addressof(keys) != this)
{
   ...
}
```
will be useful in skipping work in the one edge case.

Reply:
I've not checked, it's a straight copy of what the ctor does. I'll add the address check.